### ISSUE TITLE
Question about decimal.Decimal support

### ISSUE THREAD

**Issue by DoubleDi** [gocql/gocql#1578]
Our company is working with money and a lot of operations are done with Decimal/Numeric type.

For Go there is a great library github.com/shopspring/decimal which gives full support for marshalling, converting and math on decimal types, and is pretty well known. GoCQL uses another decimal lib called inf.Dec, and it is not as advanced as the shopspring lib. And it's a shame to remap the types from inf.Dec to decimal.Decimal and back all the time.

Would you by any change be considering of a PR of accepting support of github.com/shopspring/decimal the same what as it is done for inf.Dec?

**Comment by martin-sucha** [comment]
I don't think that we should have two dependencies for handling decimal numbers in gocql. Introducing a dependency affects every user.

You could implement [gocql.Marshaler](https://pkg.go.dev/github.com/gocql/gocql#Marshaler) and [gocql.Unmarshaler](https://pkg.go.dev/github.com/gocql/gocql#Unmarshaler) interfaces for your type in a separate package (and publish it to Github for example; I'd be happy to link to it from the README). See also [example](https://pkg.go.dev/github.com/gocql/gocql#example-package-MarshalerUnmarshaler). That type could wrap the `decimal.Decimal` and unmarshal to it directly.

Would that work for you?

**Comment by DoubleDi** [comment]
Thanks for the response.

That's definitely a way to go and we have implemented this wrapper internally. 

Just a question, but are there any possibilities to write a Marshaler/Unmarshaler, for it to be abstracted from gocql.TypeInfo?
This usecase would really help to adapt marshaling/unmarshaling of external library types to cassandra, without having the gocql library as a dependency. In precise I am talking about considering of supporting one of driver.Valuer/driver.Scanner, encoding.TextMarshaler/encoding.TextUnmarshaler or encoding.BinaryMarshaler/encoding.BinaryUnmarshaler interface for data transformation.

**Comment by martin-sucha** [comment]
The TypeInfo is required for the (un)marshaler to know how the data is encoded on the wire.
There are differences between protocol versions (e.g. collection size is int (protocol >= 3) vs smallint (protocol <= 3))
and we allow Go types depending on the target CQL type.

We can't use encoding.BinaryMarshaler because of the above, unless we restrict the type to CQL `blob`. Similarly TextMarshaler would work only with `ascii`, `text`, `varchar`.

driver.Valuer could theoretically work for some types, but:

* It seems inefficient to allocate an `interface{}` value that will then be serialized again.
* It does not solve the issue that CQL values can be serialized differently based on the protocol version.

How would a driver.Valuer implementation look like for a CQL decimal?

<hr>

So what are the issues currently that we should be solving?

For a simple scan of a row, using the wrapper that implements CQL (un)marshaling is easy.

One of the issues is that we don't want to add dependency on `github.com/shopspring/decimal` to gocql and vice-versa. The dependency is not a problem if the wrapper is published in its own module (that depends on both gocql and `github.com/shopspring/decimal`). We can link to known integrations for example from gocql README.

If the custom type (like decimal.Decimal) is embedded within a structure it can be cumbersome to put the wrapper type to all structs, in that case it might be worth implementing something like #1373 to use such integrations.

**Comment by DoubleDi** [comment]
Great explanation, thanks.

The main problem we are trying to solve is actually to avoid the creation of these wrapper types, as you need to proxy all the methods of the original methods and keep them in sync with the original type. 
example:
```
// this is a proxy model, which is marshalable to cassandra
type Decimal struct {
	decimal.Decimal
}

func (d Decimal) GreaterThanOrEqual(d2 Decimal) bool {
	return d.Decimal.GreaterThanOrEqual(d2.Decimal)
}
```

I have 2 thoughts based on your explanation:
1. Add the protocol in the configuration, when we create the connection. We will know the protocol version and know the way to (un)marshal the types.
2. Register (un)marshalers for custom types `gocql.RegisterType(decimal.Decimal{}, marshallerFunc, unmarshallerFunc)`

What do you think?

**Comment by martin-sucha** [comment]
> The main problem we are trying to solve is actually to avoid the creation of these wrapper types, as you need to proxy all the methods of the original methods and keep them in sync with the original type.

How do you scan the rows? Do you use [iter.Scan](https://pkg.go.dev/github.com/gocql/gocql#Iter.Scan) or [Scanner.Scan](https://pkg.go.dev/github.com/gocql/gocql#Scanner) directly or do you use some wrapper like scylladb/gocqlx?

If the wrapper is not embedded inside some other struct, you can use a pointer in the wrapper (just at the time of scanning). For example (for simplicity using string representation as the wire format):

```go
type DecimalMarshaler struct {
	Target *decimal.Decimal
}

func (dm DecimalMarshaler) MarshalCQL(info gocql.TypeInfo) ([]byte, error) {
	return gocql.Marshal(info, dm.Target.String())
}

func (dm *DecimalMarshaler) UnmarshalCQL(info gocql.TypeInfo, data []byte) error {
	var s string
	err = gocql.Unmarshal(info, &s)
	if err != nil {
		return err
	}
	*dm.Target = decimal.NewFromString(s)
	return nil
}

func LoadFromDB(iter *gocql.Iter) error {
	var dec decimal.Decimal
	err = iter.Scan(DecimalMarshaler{Target: &dec))
	if err != nil {
		return err
	}
	// dec is decimal.Decimal loaded from the DB now
}
```

This way it is possible to implement entirely custom unmarshaling, however if you want to support UDTs / structs you need to unmarshal recursively yourself.

<hr>

1. Your app knows the version, but driver.Valuer in external library (e.g. github.com/shopspring/decimal) does not
2. That could work, but I don't like the global state. Global state complicates testing.
3. That is too simple. TypeInfo currently can describe user defined types and collections of types,

**Comment by OleksiienkoMykyta** [comment]
Adding support for this in the future might be a good idea, but users can create wrapper-marshalers when they want to use their types, so this is not a big deal.
If comparing these two libraries:
shopspring/decimal - easier to use(more intuitive), JSON/database support(SQL integration) 
inf.Dec - custom rounding and precision control, higher-performance, low-level arithmetic
I'm on the fence about it.
@joao-r-reis, what is your take on it?

**Comment by Tang8330** [comment]
Why is this library using `*inf.Dec` and relying on another SDK?

1. The package state that it is in experimental stage
2. The package does not support converting this back out to float32/float64
3. Cassandra's official docs says that `decimal` maps to `int` or `float`. [Ref](https://cassandra.apache.org/doc/4.1/cassandra/cql/types.html#:~:text=below%20for%20details.-,decimal,Variable%2Dprecision%20decimal,-double)

Why isn't it just mapping to a float64 and if the value is an integer, it's just a float with no scale? This library is making it far more complicated than it needs to be given that Cassandra only supports up to float64.

Tagging @joao-r-reis 

<img width="1520" height="815" alt="Image" src="https://github.com/user-attachments/assets/693ef333-6e8d-4a71-bd4f-ce4e2a90fa30" />

**Comment by martin-sucha** [comment]
The documentation you linked only refers to CQL constants.

Cassandra DECIMAL has wider precision than floats. The decimal data type contains a variable precision decimal number, see the [CQL protocol documentation](https://cassandra.apache.org/doc/latest/cassandra/_attachments/native_protocol_v5.html#s5.6). Converting to float32 or float64 can lose precision, which is why gocql does not support it.

Consider a case when you have decimal data in the database and you round-trip values through gocql. Gocql avoids conversion to less precise type so that the precision of the values stored in the database is not degraded.

**Comment by Tang8330** [comment]
Ah, I see. It sounds like the same setup as Postgres: as PG supports specifying `NUMERIC(p, s)` and `NUMERIC` which the latter translates to arbitrary values.
